### PR TITLE
Support reordering between sections, support reordering validation

### PR DESCRIPTION
**Pull request by kyleve**
### Added

- [Reordering between multiple sections is now supported](https://github.com/kyleve/Listable/pull/292).
- [Introduced type safe access to `Section` content following reorder events](https://github.com/kyleve/Listable/pull/292). See `Section.filtered`.
- [`ListStateObserver.onItemReordered` was added](https://github.com/kyleve/Listable/pull/292) to observe reorder events at a list-wide level.
- [`ListLayout` was extended](https://github.com/kyleve/Listable/pull/292) to allow customization of in-progress moves. Note that `ListLayout` is not yet public.

### Changed

- [`Reordering` has been renamed to `ItemReordering`, and a new `SectionReordering` has been introduced](https://github.com/kyleve/Listable/pull/292). This allows finer-grained control over validating reorder events at both the item level and section level.
- [`ListReorderGesture` and `ItemReordering.GestureRecognizer` have been heavily refactored](https://github.com/kyleve/Listable/pull/292) to reduce visibility of internal state concerns.
- [`Item.identifier` has been renamed to `Item.anyIdentifier`](https://github.com/kyleve/Listable/pull/292). The new `Item.identifier` property is now a fully type safe identifier.
- [`ReorderingActions` was refactored](https://github.com/kyleve/Listable/pull/292) to expose less public state and ease use in UIView-backed list elements.
- [`Identifier<Represented>` is now `Identifier<Represented, Value>`; eg `Identifier<MyContent, UUID>`](https://github.com/kyleve/Listable/pull/292). This is done to support reacting to reordering events in a more type safe manner, and to make `Identifier` creation more type safe. This is a large breaking change.
- [Changed how `identifier`s for `ItemContent` are represented](https://github.com/kyleve/Listable/pull/292). `ItemContent` now returns a an identifier of a specific `IdentifierType` (eg, `String`, `UUID`, etc), which is then assembled into an `Identifier` by the containing item. Additional APIs have been added for creating `Identifier`s in a more type safe manner. This is a large breaking change.

### CODE REVIEW

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 1]
```diff
@@ -190,22 +192,49 @@ final class CollectionViewLayout : UICollectionViewLayout
         self.neededLayoutType.merge(with: context)
     }
     
+    override func invalidationContext(
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:195]: Might not need this at all?

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 2]
```diff
@@ -190,22 +192,49 @@ final class CollectionViewLayout : UICollectionViewLayout
...
+        forInteractivelyMovingItems targetIndexPaths: [IndexPath],
+        withTargetPosition targetPosition: CGPoint,
+        previousIndexPaths: [IndexPath],
+        previousPosition: CGPoint
+    ) -> UICollectionViewLayoutInvalidationContext
+    {
+        self.isReordering = true
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:202]: Centralize this onto `invalidate(with:)` via a property on the `InvalidationContext`

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 3]
```diff
@@ -190,22 +192,49 @@ final class CollectionViewLayout : UICollectionViewLayout
...
+        return context
+    }
+    
     override func invalidationContextForEndingInteractiveMovementOfItems(
         toFinalIndexPaths indexPaths: [IndexPath],
         previousIndexPaths: [IndexPath],
         movementCancelled: Bool
     ) -> UICollectionViewLayoutInvalidationContext
     {
+        isReordering = false
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:222]: Centralize this onto `invalidate(with:)` via a property on the `InvalidationContext`

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 4]
```diff
@@ -28,6 +28,8 @@ final class CollectionViewLayout : UICollectionViewLayout
         }
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:28]: In this file: Validate that we're not setting the `neededLayoutType` until the index paths of the item actually change.

Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:28]: I think this needs to loop over the moved index paths; not just the first one:

```
    override func invalidateLayout(with context: UICollectionViewLayoutInvalidationContext)
    {
        let view = self.collectionView!
        let context = context as! InvalidationContext
        
        super.invalidateLayout(with: context)
        
        // Handle Moved Items
        
        if
            let from = context.previousIndexPathsForInteractivelyMovingItems,
            let to = context.targetIndexPathsForInteractivelyMovingItems
        {
            let from = from[0]
            let to = to[0]
```

Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:28]: Can I remove the `liveIndexPath` stuff?

**Comment by kyleve** [comment]
Still crashes:
- Go to reordering demo
- Move one item from last section to first section
- Drag any item from first section down to bottom of last section (don't let it go)
- Move it up towards the first section
- Crash (collection view tries to look up (1,1))

File: ListableUI/Sources/Identifier.swift [hunk 5]
```diff
@@ -77,6 +77,10 @@ public final class Identifier<Represented> : AnyIdentifier
 ///
 public class AnyIdentifier : Hashable, CustomDebugStringConvertible
 {
+    public var base : Any {
```
Comment (kyleve) [inline ListableUI/Sources/Identifier.swift:80]: Need to comment this

File: ListableUI/Sources/HeaderFooter/HeaderFooterContent.swift [hunk 6]
```diff
@@ -128,6 +128,8 @@ public protocol HeaderFooterContent
 
 /// Information about the current state of the content, which is passed to `apply(to:for:with:)`
 /// during configuration and preparation for display.
+///
+/// TODO: Rename to `ApplyHeaderFooterContext`
```
Comment (kyleve) [inline ListableUI/Sources/HeaderFooter/HeaderFooterContent.swift:132]: Do this by renaming, then deprecate the old name

File: ListableUI/Sources/Internal/Presentation State/PresentationState.ItemState.swift [hunk 7]
```diff
@@ -461,9 +468,37 @@ extension PresentationState
             }
         }
         
-        func moved(with result : Reordering.Result)
+        func beginReorder(with environment: ListEnvironment) {
```
Comment (kyleve) [inline ListableUI/Sources/Internal/Presentation State/PresentationState.ItemState.swift:471]: Document this and the below

File: ListableUI/Sources/Internal/Presentation State/PresentationState.ItemState.swift [hunk 8]
```diff
@@ -461,9 +468,37 @@ extension PresentationState
...
+            }
+            
+            self.isReordering = false
+            
+            UIView.animate(withDuration: 0.15) {
+                self.applyToVisibleCell(with: environment)
+            }
+        }
+        
+        private(set) var isReordering : Bool = false
```
Comment (kyleve) [inline ListableUI/Sources/Internal/Presentation State/PresentationState.ItemState.swift:497]: Make sure this only runs if things are NOT cancelled

File: ListableUI/Sources/Item/ItemContent.swift [hunk 9]
```diff
@@ -226,7 +228,11 @@ public struct ApplyItemContentInfo
     public var position : ItemPosition
     
     /// Provides access to actions to handle re-ordering the content within the list.
-    public var reordering : ReorderingActions
+    public var reorderingActions : ReorderingActions
+    
+    /// If the item can be reordered.
+    /// Use this property to determine in your `ItemContent` should render a reorder control.
```
Comment (kyleve) [inline ListableUI/Sources/Item/ItemContent.swift:234]: `in your` -> `if your`

File: ListableUI/Sources/Item/Reordering.swift [hunk 10]
```diff
@@ -8,37 +8,225 @@
...
+            case .possible: break
+            case .began:
+                let canStart = self.onStart?()
+                
+                let center = self.view?.firstSuperview(ofType: UICollectionViewCell.self)?.center
+                
+                if let center = center, canStart == true {
+                    self.initialCenter = center
+                } else {
+                    self.state = .cancelled
```
Comment (kyleve) [inline ListableUI/Sources/Item/Reordering.swift:168]: Do I need to do `onEnd`, etc, here?

File: ListableUI/Sources/Item/Reordering.swift [hunk 11]
```diff
@@ -8,37 +8,225 @@
...
+            @unknown default: listableFatal()
+            }
+        }
+    }
+}
+
+
+extension Reordering.Sections {
+    
+    func canMove(from : PresentationState.SectionState, to : PresentationState.SectionState) -> Bool {
```
Comment (kyleve) [inline ListableUI/Sources/Item/Reordering.swift:190]: Test

File: ListableUI/Sources/Item/Reordering.swift [hunk 12]
```diff
@@ -8,37 +8,225 @@
...
+        case .specific(let IDs):
+            return IDs.contains(to.model.identifier)
+        }
+    }
+}
+
+
+extension Reordering {
+    
+    func destination(
```
Comment (kyleve) [inline ListableUI/Sources/Item/Reordering.swift:208]: Test

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 13]
```diff
@@ -163,21 +165,23 @@ final class CollectionViewLayout : UICollectionViewLayout
...
+                if info.from != info.to {
+                    self.layout.content.move(from: info.from, to: info.to)
+                    self.layout.content.reindexIndexPaths()
+                }
+
+            case .complete(_):
+                break
+
+            case .cancelled(let info):
+                self.layout.content.move(from: info.from, to: info.to)
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:184]: Test this manually

File: ListableUI/Sources/Layout/ListLayout/ListLayout.swift [hunk 14]
```diff
@@ -79,6 +79,17 @@ public protocol AnyListLayout : AnyObject
     )
     
     func setZIndexes()
+    
+    //
+    // MARK: Configuring Reordering
+    //
+    
+    func adjust(
```
Comment (kyleve) [inline ListableUI/Sources/Layout/ListLayout/ListLayout.swift:87]: Instead of using `UICollectionViewLayoutAttributes` directly here; let's add a struct type that only exposes the props needed.

File: ListableUI/Sources/Layout/ListLayout/ListLayoutContent.swift [hunk 15]
```diff
@@ -198,6 +189,39 @@ public final class ListLayoutContent
         self.sections[to.section].items.insert(info, at: to.item)
     }
     
+    func move(from : [IndexPath], to : [IndexPath])
```
Comment (kyleve) [inline ListableUI/Sources/Layout/ListLayout/ListLayoutContent.swift:192]: Test

File: ListableUI/Sources/ListView/ListView.DataSource.swift [hunk 16]
```diff
@@ -75,27 +77,34 @@ internal extension ListView
...
         }
         
         func collectionView(
             _ collectionView: UICollectionView,
-            moveItemAt sourceIndexPath: IndexPath,
-            to destinationIndexPath: IndexPath
+            moveItemAt from: IndexPath,
+            to: IndexPath
+        ) {
+            self.storage.moveItem(from: from, to: to)
```
Comment (kyleve) [inline ListableUI/Sources/ListView/ListView.DataSource.swift:95]: Verify: This is only called when the indexes are actually different, right? Maybe throw in a guard check to be safe against regressions...

File: ListableUI/Sources/ListView/ListView.Delegate.swift [hunk 17]
```diff
@@ -180,24 +181,52 @@ extension ListView
...
+            let item = self.layoutManager.stateFor(itemAt: from)
+            
+            guard let reordering = item.anyModel.reordering else {
+                return from
+            }
+            
+            let fromSection = self.presentationState.sections[from.section]
+            let toSection = self.presentationState.sections[to.section]
+            
+            return reordering.destination(
```
Comment (kyleve) [inline ListableUI/Sources/ListView/ListView.Delegate.swift:224]: Todo: This does not move the item from the from section and pass it to the to section. Should we?

File: ListableUI/Sources/ReorderingActions.swift [hunk 18]
```diff
@@ -18,47 +18,63 @@ public final class ReorderingActions
...
-    public func beginMoving() -> Bool
+    public func start() -> Bool
     {
+        guard let item = self.item else {
+            return false
+        }
+        
         guard self.isMoving == false else {
-            return true
+            return false
```
Comment (kyleve) [inline ListableUI/Sources/ReorderingActions.swift:28]: Maybe should still return true?

File: ListableUI/Sources/ReorderingActions.swift [hunk 19]
```diff
@@ -18,47 +18,63 @@ public final class ReorderingActions
...
-        self.delegate?.updateInteractiveMovementTargetPosition(with: recognizer)
+        guard let item = self.item else {
+            return
+        }
+        
+        self.delegate?.updateReorderTargetPosition(with: recognizer, for: item)
     }
     
-    public func end()
+    public func end(_ cancelled : Bool)
```
Comment (kyleve) [inline ListableUI/Sources/ReorderingActions.swift:57]: Consider making this an enum instead of a bool; less confusing

File: ListableUI/Sources/ReorderingActions.swift [hunk 20]
```diff
@@ -18,47 +18,63 @@ public final class ReorderingActions
...
         }
         
+        guard let item = self.item else {
+            return
+        }
+        
         self.isMoving = false
         
-        self.delegate?.endInteractiveMovement()
+        self.delegate?.endReorder(for: item)
```
Comment (kyleve) [inline ListableUI/Sources/ReorderingActions.swift:69]: Shouldn't this call cancel in the cancel case?

File: ListableUI/Sources/ReorderingActions.swift [hunk 21]
```diff
@@ -18,47 +18,63 @@ public final class ReorderingActions
...
     }
 }
 
 
 protocol ReorderingActionsDelegate : AnyObject
```
Comment (kyleve) [inline ListableUI/Sources/ReorderingActions.swift:74]: Merge end vs cancelled into one method with a param to specify the result?

File: ListableUI/Sources/Layout/CollectionViewLayout.swift [hunk 22]
```diff
@@ -163,21 +165,23 @@ final class CollectionViewLayout : UICollectionViewLayout
...
             
-            if from != to {
-                context.performedInteractiveMove = true
-                self.layout.content.reindexLiveIndexPaths()
+            switch action {
+            case .inProgress(let info):
+                
+                if info.from != info.to {
+                    self.layout.content.move(from: info.from, to: info.to)
+                    self.layout.content.reindexIndexPaths()
```
Comment (kyleve) [inline ListableUI/Sources/Layout/CollectionViewLayout.swift:177]: Move this into the move method

File: ListableUI/Sources/Layout/CollectionViewLayout.swift (hunk 13 above)
Comment (kyleve) [reply ListableUI/Sources/Layout/CollectionViewLayout.swift:184]: How can I test this... automated...ly?

File: ListableUI/Sources/ListView/ListView.Delegate.swift (hunk 17 above)
Comment (kyleve) [reply ListableUI/Sources/ListView/ListView.Delegate.swift:224]: I don't think it should. Going to leave as is; since the change has not been committed.

File: ListableUI/Sources/ListView/ListView.LayoutManager.swift [hunk 23]
```diff
@@ -22,6 +22,10 @@ extension ListView
             self.collectionView = collectionView
         }
         
+        func stateFor(itemAt indexPath: IndexPath) -> AnyPresentationItemState {
```
Comment (kyleve) [inline ListableUI/Sources/ListView/ListView.LayoutManager.swift:25]: `stateForItem(at:`

Comment (kyleve) [reply ListableUI/Sources/ListView/ListView.LayoutManager.swift:25]: Done

File: Demo/Sources/Demos/Demo Screens/ReorderingViewController.swift [hunk 24]
```diff
@@ -14,95 +14,83 @@ import BlueprintUI
...
 
 
-final class ReorderingViewController : UIViewController
+final class ReorderingViewController : ListViewController
 {
-    let list = ListView()
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Reload", style: .plain, target: self, action: #selector(reload))
```
Comment (kyleve) [inline Demo/Sources/Demos/Demo Screens/ReorderingViewController.swift:22]: Still need to test this.

Comment (kyleve) [reply Demo/Sources/Demos/Demo Screens/ReorderingViewController.swift:22]: This doesn't work; but since its an edge case; I'm going to land and work on getting into SPOS. Can follow up w/ this.

File: ListableUI/Sources/Layout/ListLayout/ListContentLayoutAttributes.swift [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+            
+            self.size = newValue.size
+        }
+    }
+    
+    //
+    // MARK: Transforms & Layout
+    //
+    
+    public var transform: CGAffineTransform
```
Comment (kyleve) [inline ListableUI/Sources/Layout/ListLayout/ListContentLayoutAttributes.swift:53]: Consider:

```
    public enum Transform {
        case affine(CGAffineTransform)
        case transform3d(CATransform3D)
    }
```

File: ListableUI/Sources/ListView/ListView.CollectionViewChanges.swift [hunk 26]
```diff
@@ -25,6 +25,7 @@ internal extension ListView
         var movedItems : [ItemChanges.Moved] = []
                 
         var hasIndexAffectingChanges : Bool {
+            return true
```
Comment (kyleve) [inline ListableUI/Sources/ListView/ListView.CollectionViewChanges.swift:28]: - [ ] Revert before merge

File: ListableUI/Sources/Identifier.swift [hunk 27]
```diff
@@ -14,36 +14,59 @@ import Foundation
...
+    /// strongly typed identifiers with the correct `Value` type:
+    /// ```swift
+    /// MyItem.identifier(with: "my-id")  // ✅ OK
+    /// MyItem.identifier(with: 1)        // 🚫 Error: MyItem's IdentifierValue is String.
+    /// ```
+    ///
+    /// You can also read  ``Item.identifier-swift.property`` or ``AnyItem.anyIdentifier-swift.property``
+    /// to get the identifier of an item that has been created in a type safe manner.
+    ///
+    init(_ value : Value)
```
Comment (alanjzhang) [inline ListableUI/Sources/Identifier.swift:69]: I think this needs to be public - here's an [example](https://github.com/squareup/ios-register/blob/aba13e6140ad1e22ad6b538a2b31d35744f304f2/Frameworks/SquareSettings/Impl/Sources/Master/SettingsMasterScreen.swift#L140) of where this is needed. I just pointed my local Podfile to your branch and had to update a few things from the breaking API changes.

Comment (alanjzhang) [inline ListableUI/Sources/Identifier.swift:69]: I think this needs to be public. I just pointed my local Podfile to your branch and had to update a few things from the breaking API changes.

File: ListableUI/Sources/Layout/Table/TableListLayout.swift [hunk 28]
```diff
@@ -398,7 +398,7 @@ final class TableListLayout : ListLayout
     typealias LayoutAppearance = TableAppearance
     
     static var defaults: ListLayoutDefaults {
-        .init(itemInsertAndRemoveAnimations: .top)
+        .init(itemInsertAndRemoveAnimations: .fade)
```
Comment (kyleve) [inline ListableUI/Sources/Layout/Table/TableListLayout.swift:401]: Changing this default; it looks better.
